taichi/transforms/make_thread_local.cpp

```diff
@@ -0,0 +1,153 @@
+#include "taichi/ir/ir.h"
+#include "taichi/ir/transforms.h"
+#include "taichi/ir/analysis.h"
+#include "taichi/ir/visitors.h"
+
+TLANG_NAMESPACE_BEGIN
+
+namespace {
+
+bool is_atomic_op_linear(AtomicOpType op_type) {
+  return op_type == AtomicOpType::add || op_type == AtomicOpType::sub;
+}
+
+void make_thread_local_offload(OffloadedStmt *offload) {
+  // TODO: deal with struct for
+  if (offload->task_type != offload->range_for)
+    return;
+
+  // Gather all atomic adds/subs destinations
+  // We use std::vector instead of std::set to keep an deterministic order here.
+  std::vector<GlobalPtrStmt *> atomic_destinations;
+  // TODO: this is again an abuse since it gathers nothing. Need to design a IR
+  // map/reduce system
+  auto linear_atomics =
+      irpass::analysis::gather_statements(offload, [&](Stmt *stmt) {
+        if (auto atomic_op = stmt->cast<AtomicOpStmt>()) {
+          if (is_atomic_op_linear(atomic_op->op_type)) {
+            // Local atomics does not count
```

Comment:
```suggestion
            // Local or global tmp atomics does not count
```
I don't think global tmps are supported in this PR?